dashboard/app/models/user.rb

```diff
@@ -804,12 +802,20 @@ def self.new_with_session(params, session)
     end
   end
 
+  def managing_own_account?
+    provider.blank? || (provider == User::PROVIDER_MANUAL)
+  end
+
   def password_required?
-    # password is required if:
-    (!persisted? || # you are a new user
-     !password.nil? || !password_confirmation.nil?) && # or changing your password
-      (provider.blank? || (User::PROVIDER_MANUAL == provider)) # and you are a person creating your own account
-    # (as opposed to a person who had their account created for them or are logging in with oauth)
+    # Password is not required if the user is not managing their own account
+    # (i.e., someone is creating their account for them or the user is using OAuth).
+    return false unless managing_own_account?
+
+    # Password is required for:
+    # New users with no encrypted_password set and users changing their password.
+    new_without_password = !persisted? && encrypted_password.blank?
+    is_changing_password = password.present? || password_confirmation.present?
```

Comment:
Does this code ever run in an non-password-change scenario?  I thought we could have `password_confirmation` present in other situations, for example when a user is updating their email address.

The logic might be correct, but maybe the variable name is misleading.

Reply:
that's a great question. this validation currently runs on every save, so it would run when a user is updating their email address. however, `password` and `password_confirmation` are always `nil` on the User model unless those properties have changed and not yet been saved. on save, those properties are set to `nil` again and `encrypted_password` is updated if validations succeed. 

the current implementation of this method is never taking other properties changing into account--that's being handled manually in controller actions via [`needs_password?`](https://github.com/code-dot-org/code-dot-org/blob/496286a5562bbf1731f3f1588febd78d131d0341/dashboard/app/controllers/registrations_controller.rb#L319-L334) (but should definitely be in model validations).

Reply:
another way of saying this that is more succinct and probably makes more sense is that the new implementation matches the current implementation in this regard (i just refactored the setup to be--hopefully--more readable)